### PR TITLE
Remove defer from 3D Tiles, glTF, and point cloud loading

### PR DESCRIPTION
**Pull request by ggetz**
Part of https://github.com/CesiumGS/cesium/issues/10178
Fixes https://github.com/CesiumGS/cesium/issues/10363

Removed use of `defer` from 3D Tiles, glTF, and point cloud loading.

I don't think this is the be-all-end-all improvement, since in a few places we are still not entirely encapsulating the entire promise resolve/reject behavior within a `new Promise` callback, and probably won't be able to do so without a good amount more refactoring, and probably deprecating `readyPromises` throughout.

But, I think this should be at least a step in the right direction as this does remove `defer` from plus this does now handle several errors which were not being handled properly before. One particular area this cleaned up was handling 3D Tiles promises when destroyed. In doing so, this allow us to re-introduce all of the specs that had to be excluded as part of https://github.com/CesiumGS/cesium/pull/10149.

TODO:
- [x] Merge main
- [x] Confirm performance of loading was not affected
- [x] Clear up https://github.com/CesiumGS/cesium/compare/defer-b-gone?expand=1#diff-70f07d262fd26759318ee0562b7a1a45915ea04e2d091f654a41018ecb504a92R113
- [x] Update `CHANGES.md`

### CODE REVIEW

File: Source/Scene/BufferLoader.js [hunk 1]
```diff
@@ -98,32 +97,39 @@ Object.defineProperties(BufferLoader.prototype, {
...
 
 function loadExternalBuffer(bufferLoader) {
   const resource = bufferLoader._resource;
   bufferLoader._state = ResourceLoaderState.LOADING;
-  BufferLoader._fetchArrayBuffer(resource)
+  const promise = BufferLoader._fetchArrayBuffer(resource);
+
+  // TODO: I don't think this is OK-- The request is cancelled, no?
+  if (!defined(promise)) {
+    return Promise.resolve(bufferLoader);
```
Comment (ggetz) [inline Source/Scene/BufferLoader.js:114]: This was one area I wasn't sure about, and could possible be an existing bug. It's possible for `BufferLoader._fetchArrayBuffer` to return `undefined` if the request scheduler is full, meaning the request needs to be retried later. But the existing behavior appeared to swallow the error.

Comment (lilleyse) [reply Source/Scene/BufferLoader.js:114]: `resource.throttle` should be false here so the promise should always be defined. Only the tile's main content (.glb or .gltf) is throttled currently. I edited one of the test tilesets to have an external bin file and confirmed that this is the case.

Comment (lilleyse) [reply Source/Scene/BufferLoader.js:114]: No, that issue is about improving throttling behavior for the things that are being throttled.

**Comment by lilleyse** [comment]
Just to confirm, does this also fix https://github.com/CesiumGS/cesium/issues/10222?

**Comment by ggetz** [comment]
> Just to confirm, does this also fix https://github.com/CesiumGS/cesium/issues/10222?

@lilleyse No. That will need changes in `Viewer.js`, and I'm aiming to keep this PR as small as possible.

**Comment by ggetz** [comment]
@lilleyse This is ready for review. Could you take a look, or recommend a reviewer familiar with glTF and 3D Tile loading?

**Review by lilleyse — COMMENTED** [review]
Thanks @ggetz this looks pretty good. It took me a little while to understand the `process` functions refactor but I see why it's done that way.

I wish we had avoided `defer` in the first place and saved you all this trouble :sweat_smile:

**Review by ptrgags — COMMENTED** [review]
@ggetz I took a look at `Cesium3DTile` and `ModelExperimental`-related code. The approach looks good, just a couple comments from me.

File: Source/Scene/Cesium3DTile.js [hunk 2]
```diff
@@ -1189,69 +1190,76 @@ function requestSingleContent(tile) {
...
   const tileset = tile._tileset;
   tile._contentState = Cesium3DTileContentState.LOADING;
-  tile._contentReadyToProcessPromise = defer();
-  tile._contentReadyPromise = defer();
   ++tileset.statistics.numberOfPendingRequests;
+  const contentReadyToProcessPromise = promise.then(function (arrayBuffer) {
+    if (tile.isDestroyed()) {
+      // Tile is unloaded before the content finishes loading
+      singleContentFailed(tile, tileset);
+      return;
```
[the remark this answers is not in the document]
Comment (ggetz) [reply Source/Scene/Cesium3DTile.js:1200]: I actually ended up matching `multipleContentFailed` to `singleContentFailed` to be more explicit about what is being returned in turn of `undefined` vs a rejection.

File: Source/Scene/GltfLoader.js [hunk 3]
```diff
@@ -231,40 +235,130 @@ GltfLoader.prototype.load = function () {
   this._textureState = GltfLoaderState.LOADING;
 
   const that = this;
-  gltfJsonLoader.promise
+  let textureProcessPromise;
+  const processPromise = new Promise(function (resolve, reject) {
+    textureProcessPromise = new Promise(function (
```
[the remark this answers is not in the document]
Comment (ggetz) [reply Source/Scene/GltfLoader.js:246]: I don't think this is possible. The `process` function needs access to the `resolve` and `reject` callbacks for both the master readyPromise and texture readyPromise. If we move the callbacks to member functions, then we just go back to the problematic `defer` pattern.

**Comment by ggetz** [comment]
Thanks @lilleyse and @ptrgags! Updated!

**Comment by lilleyse** [comment]
Latest changes look good. Thanks @ggetz.

**Pull request merged**